Firestore/core/src/firebase/firestore/core/firestore_client.cc

```diff
@@ -248,30 +248,57 @@ void FirestoreClient::EnableNetwork(StatusCallback callback) {
   });
 }
 
-void FirestoreClient::Terminate(StatusCallback callback) {
+void FirestoreClient::TerminateAsync(StatusCallback callback) {
   auto shared_this = shared_from_this();
   worker_queue()->EnqueueAndInitiateShutdown([shared_this, callback] {
-    shared_this->credentials_provider_->SetCredentialChangeListener(nullptr);
+    shared_this->TerminateInternal();
 
-    // If we've scheduled LRU garbage collection, cancel it.
-    if (shared_this->lru_callback_) {
-      shared_this->lru_callback_.Cancel();
-    }
-    shared_this->remote_store_->Shutdown();
-    shared_this->persistence_->Shutdown();
-  });
-
-  // This separate enqueue ensures if `terminate` is called multiple times
-  // every time the callback is triggered. If it is in the above
-  // enqueue, it might not get executed because after first `terminate`
-  // all operations are not executed.
-  worker_queue()->EnqueueEvenAfterShutdown([shared_this, callback] {
     if (callback) {
       shared_this->user_executor()->Execute([=] { callback(Status::OK()); });
     }
   });
 }
 
+void FirestoreClient::Terminate() {
+  std::mutex mutex;
+  bool terminated = false;
+  std::condition_variable terminated_condition;
+
+  std::unique_lock<std::mutex> outer_lock{mutex};
+
+  LOG_DEBUG("Terminate: starting");
```

Comment:
Question: is this temporary logging, or do you intend to keep it?

Reply:
This was a mistake; removed.